### PR TITLE
[CLI-323] Added Supplier<T> defaults for getParsedOptionValue

### PR DESCRIPTION
**Pull request by Claudenw**
Fix for CLI-323 https://issues.apache.org/jira/browse/CLI-323

### CODE REVIEW

**Review by garydgregory — CHANGES_REQUESTED** [review]
Hi @Claudenw 
Thank you for your PR.
You'll need to add tests to cover the new feature.

**Comment by Claudenw** [comment]
I think I have this fixed now.

File: src/main/java/org/apache/commons/cli/CommandLine.java [hunk 1]
```diff
@@ -423,7 +438,7 @@ public <T> T getParsedOptionValue(final char opt, final T defaultValue) throws P
      * @since 1.5.0
      */
     public <T> T getParsedOptionValue(final Option option) throws ParseException {
-        return  getParsedOptionValue(option, null);
+        return  getParsedOptionValue(option, ()-> null);
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/cli/CommandLine.java:441]: Missing a space after `()`.

File: src/main/java/org/apache/commons/cli/CommandLine.java [hunk 2]
```diff
@@ -437,15 +452,27 @@ public <T> T getParsedOptionValue(final Option option) throws ParseException {
...
+     * @see PatternOptionBuilder
+     * @since 1.7.0
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getParsedOptionValue(final Option option, final Supplier<T> defaultValue) throws ParseException {
+        final String res = option == null ? null : getOptionValue(option);
 
         try {
-            return res == null ? defaultValue : (T) option.getConverter().apply(res);
+            return res == null ? defaultValue.get() : (T) option.getConverter().apply(res);
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/cli/CommandLine.java:475]: Hi @Claudenw 
You're missing a test for when `defaultValue` is null.

File: src/test/java/org/apache/commons/cli/CommandLineTest.java [hunk 3]
```diff
@@ -166,7 +166,27 @@ public void testGetParsedOptionValueWithOption() throws Exception {
...
+        final Option optF = Option.builder("f").hasArg().build();
+        options.addOption(optI);
+        options.addOption(optF);
+
+        final CommandLineParser parser = new DefaultParser();
+        final CommandLine cmd = parser.parse(options, new String[] {"-i", "123"});
+
+        assertEquals(123, ((Number) cmd.getParsedOptionValue(optI)).intValue());
+        assertEquals("foo", cmd.getParsedOptionValue(optF, "foo"));
+        assertEquals("foo", cmd.getParsedOptionValue(optF, ()-> "foo"));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/cli/CommandLineTest.java:181]: `() -> ...`

**Review by garydgregory — CHANGES_REQUESTED** [review]

File: src/test/java/org/apache/commons/cli/CommandLineTest.java (hunk 3 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/cli/CommandLineTest.java:181]: If you know how to fix the checkstyle to look for this error I would appreciate it.

Comment (garydgregory) [reply src/test/java/org/apache/commons/cli/CommandLineTest.java]: > If you know how to fix the checkstyle to look for this error I would appreciate it.

Hi @Claudenw 
Done. Please rebase on git master.

File: src/test/java/org/apache/commons/cli/CommandLineTest.java [hunk 4]
```diff
@@ -166,7 +179,27 @@ public void testGetParsedOptionValueWithOption() throws Exception {
...
+        final CommandLineParser parser = new DefaultParser();
+        final CommandLine cmd = parser.parse(options, new String[] {"-i", "123"});
+
+        assertEquals(123, ((Number) cmd.getParsedOptionValue(optI)).intValue());
+        assertEquals("foo", cmd.getParsedOptionValue(optF, "foo"));
+        assertEquals("foo", cmd.getParsedOptionValue(optF, () -> "foo"));
+        assertEquals("foo", cmd.getParsedOptionValue("f", "foo"));
+        assertEquals("foo", cmd.getParsedOptionValue("f", () -> "foo"));
+        assertEquals("foo", cmd.getParsedOptionValue('f', "foo"));
+        assertEquals("foo", cmd.getParsedOptionValue('f', () -> "foo"));
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/cli/CommandLineTest.java:198]: The above checks are good but are all "happy paths".
I think you need assertions calling `getParsedOptionValue()` with a null Supplier and a Supplier that returns null since the runtime now supports this.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@Claudenw 
TY for your updates. Please see my comment.

File: src/test/java/org/apache/commons/cli/CommandLineTest.java (hunk 4 above)
Comment (Claudenw) [reply src/test/java/org/apache/commons/cli/CommandLineTest.java:195]: Added tests.
